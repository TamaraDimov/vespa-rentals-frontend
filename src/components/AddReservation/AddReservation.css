.reservation-container {
  border: 0.5px solid #ffffff94;
  background: url(../../assets/Reserve\ bg.jpg);
  background-color: #e3a20a53;
  background-size: cover;
  background-attachment: fixed;
  background-repeat: no-repeat;
  width: 100%;
  height: 100vh;
}

.reservation-form {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  width: 60vw;
  margin: 10% auto;
  padding: 3%;
  background: #230b0093;
  border: 2px solid #de4a00;
  border-radius: 9px;
}

.reservation-form label {
  font-size: calc(0.6rem + 1vw);
  font-weight: bold;
  color: white;
  line-height: 2rem;
  width: 100%;
  align-items: center;
  justify-content: center;
}

.reservation-form input,
.reservation-form select {
  width: 100%;
  font-size: calc(0.6rem + 1vw);
  margin-bottom: 10px;
  border: 1px solid #ccc;
  border-radius: 7px;
  color: #fff;
  outline: none;
  background: rgba(224, 219, 219, 0.245);
  font-family: 'Trebuchet MS', sans-serif;
}

.reservation-form input:focus,
.reservation-form select:focus {
  outline: none;
  box-shadow: 0 0 0 2px #ff8846;
  color: #000;
}

.reservation-form button {
  color: #fff;
  background-color: #e35e0a;
  font-weight: 700;
  height: 60px;
  padding: 10px 30px;
  width: 80%;
  border-radius: 40px;
  border: none;
  text-transform: uppercase;
  font-size: 16px;
  letter-spacing: 1.3px;
  -webkit-transition: 0.2s all;
  transition: 0.2s all;
  margin: 5% auto;
}

/* Color options */

.reservation-form button:hover {
  color: #b23000;
  background-color: #fff;
}

/* Layout and UI */

@media (min-width: 768px) {
  .reservation-form {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    width: 50vw;
    margin: 5% auto;
    padding: 6%;
    background: #230b0093;
    border: 2px solid #de4a00;
    border-radius: 9px;
  }

  .reservation-container {
    flex-direction: row;
    flex-wrap: wrap;
  }

  .form-label {
    flex: 1;
    margin-right: 1rem;
  }

  .form-input,
  .form-select {
    width: 50%;
  }

  .form-select#motorcycle_id {
    width: 50%;
  }

  button {
    width: auto;
  }
}

@media (min-width: 1200px) {
  .reservation-form {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    width: 40vw;
    margin: auto;
    padding: 2rem;
    background: #230b0093;
    border: 2px solid #de4a00;
    border-radius: 9px;
  }

  .reservation-form label {
    font-size: calc(0.4rem + 1vw);
    font-weight: bold;
    color: white;
    line-height: 2rem;
    width: 100%;
    align-items: center;
    justify-content: center;
  }

  .reservation-form input,
  .reservation-form select {
    width: 100%;
    font-size: calc(0.4rem + 1vw);
    padding: calc(0.3rem + 1vw);
    margin-bottom: 10px;
    border: 1px solid #ccc;
    border-radius: 7px;
    color: #fff;
    outline: none;
    background: rgba(224, 219, 219, 0.245);
    font-family: 'Trebuchet MS', sans-serif;
  }

  .form-input,
  .form-select {
    width: 100%;
    margin-top: 2%;
  }

  .form-select#motorcycle_id {
    width: 100%;
  }
}

/* Other styles */

.form-input {
  padding: calc(0.3rem + 1vw);
  font-size: 1rem;
}

.form-select {
  padding: 0.5rem;
  font-size: 1rem;
}

@media only screen and (max-width: 768px) {
  .reservation-form {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    width: 70vw;
    margin: 15% auto;
    padding: 5%;
    background: #230b0093;
    border: 2px solid #de4a00;
    border-radius: 9px;
  }

  .form-input {
    padding: calc(0.2rem + 1vw);
    font-size: 1rem;
  }

  .form-select {
    padding: 0.5rem;
    font-size: 1rem;
  }
}
