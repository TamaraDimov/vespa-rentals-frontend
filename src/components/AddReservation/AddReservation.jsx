import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { toast } from 'react-toastify';
import { addReservation } from '../../redux/reservationActions';
import { fetchMotorcycle } from '../../redux/reducers/motorcycleSlice';
import './AddReservation.css';

function AddReservation() {
  const dispatch = useDispatch();
  const status = useSelector((state) => state.reservation.status);
  const user = useSelector((state) => state.user.user);
  const token = user ? user.user.token : '';
  const [reservation, setReservation] = useState({
    start_date: '',
    end_date: '',
    city: '',
    user_id: user && user.user.data ? user.user.data.id : '',
    motorcycle_id: '',
  });
  const motorcycle = useSelector((state) => state.motorcycle);

  useEffect(() => {
    dispatch(fetchMotorcycle(token));
  }, [dispatch, token]);

  const handleChange = (event) => {
    const { name, value } = event.target;
    setReservation((prevState) => ({
      ...prevState,
      [name]: value,
    }));
  };

  const handleSubmit = (event) => {
    event.preventDefault();
    try {
      dispatch(addReservation(reservation));
      setTimeout(() => {
        setReservation({
          start_date: '',
          end_date: '',
          city: '',
          user_id: user && user.user.data ? user.user.data.id : '',
          motorcycle_id: '',
        });
      }, 3000);
    } catch (error) {
      toast.success('Failed to add reservation. Please try again later.');
    }
  };

  const cities = [
    'New York',
    'Los Angeles',
    'Chicago',
    'Houston',
    'Philadelphia',
  ];

  return (
    <div className="reservation-container">
      <form onSubmit={handleSubmit} className="reservation-form">
        <label htmlFor="start_date" className="form-label">
          Start Date:
          <input
            type="date"
            name="start_date"
            id="start_date"
            value={reservation.start_date}
            onChange={handleChange}
            className="form-input"
          />
        </label>
        <label htmlFor="end_date" className="form-label">
          End Date:
          <input
            type="date"
            name="end_date"
            id="end_date"
            value={reservation.end_date}
            onChange={handleChange}
            className="form-input"
          />
        </label>
        <label htmlFor="city" className="form-label">
          City:
          <select
            name="city"
            id="city"
            value={reservation.city}
            onChange={handleChange}
            className="form-select"
          >
            <option value="">Select a city</option>
            {cities.map((city) => (
              <option key={city} value={city}>
                {city}
              </option>
            ))}
          </select>
        </label>
        <label htmlFor="motorcycle_id" className="form-label">
          Motorcycle:
          <select
            name="motorcycle_id"
            id="motorcycle_id"
            value={reservation.motorcycle_id}
            onChange={handleChange}
            className="form-select"
          >
            <option value="">Select a motorcycle</option>
            {Array.isArray(motorcycle.motorcycle)
            && motorcycle.motorcycle.map((motor) => (
              <option key={motor.id} value={motor.id}>
                {motor.name}
              </option>
            ))}
          </select>
        </label>
        <button type="submit" disabled={status === 'saving'}>
          {status === 'saving' ? 'Saving...' : 'Add Reservation'}
        </button>
      </form>
    </div>
  );
}

export default AddReservation;
