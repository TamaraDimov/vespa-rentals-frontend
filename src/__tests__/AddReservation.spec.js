import '@testing-library/jest-dom';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import AddReservation from '../components/AddReservation/AddReservation';
import 'mock-local-storage';

const mockStore = configureMockStore();
const testuser = {
  user: {
    data: {
      id: 1,
      email: 'kiflekifle23@gmail.com',
      created_at: '2023-09-22T13:51:50.089Z',
    },
    token:
      'eyJhbGciOiJIUzI1NiJ9.eyJqdGkiOiJlYjI2ZTA2Yy1lMTM1LTRmMmEtYTYwZi1iYmMwM2I1MGE2MmQiLCJzdWIiOiIxIiwic2NwIjoidXNlciIsImF1ZCI6bnVsbCwiaWF0IjoxNjk2MjI2NTUwLCJleHAiOjE3MDE0MTA1NTB9.g0XLlQ7sc-1r0KIevNXVKwwH61FzdXM_sB-tpS6WWK4',
  },
};
const store = mockStore({
  motorcycle: {
    motorcycle: [
      {
        id: 1,
        name: 'Honda CBR600RR',
        model: 'CBR600RR',
        photo: 'https://example.com/photo1.jpg',
        description: 'Description for Honda CBR600RR',
      },
      {
        id: 2,
        name: 'Yamaha YZF-R6',
        model: 'YZF-R6',
        photo: 'https://example.com/photo2.jpg',
        description: 'Description for Yamaha YZF-R6',
      },
    ],
  },
  reservation: {
    reservation: [
      {
        id: 1, start_date: '2023-10-01', end_date: '2023-10-02', city: 'Chicago', user_id: 1, motorcycle_id: 1,
      },
    ],
  },
  user: {
    isLoading: false,
    user: testuser,
    token: testuser.token,
  },
});
describe('AddReservation', () => {
  let mockLocalStorage;
  let originalLocalStorage;

  beforeAll(() => {
    originalLocalStorage = global.localStorage;

    mockLocalStorage = {
      getItem: jest.fn(),
      setItem: jest.fn(),
      removeItem: jest.fn(),
      clear: jest.fn(),
    };

    const localStorageMock = {
      ...mockLocalStorage,
      getItem: jest.fn().mockImplementation((key) => mockLocalStorage[key]),
      setItem: jest.fn().mockImplementation((key, value) => {
        mockLocalStorage[key] = value;
      }),
      removeItem: jest.fn().mockImplementation((key) => {
        delete mockLocalStorage[key];
      }),
      clear: jest.fn().mockImplementation(() => {
        mockLocalStorage = {};
      }),
    };

    Object.defineProperty(global, 'localStorage', {
      value: localStorageMock,
    });

    localStorageMock.getItem.mockImplementation(() => JSON.stringify(testuser));
  });

  afterAll(() => {
    Object.defineProperty(global, 'localStorage', {
      value: originalLocalStorage,
    });
  });

  let addReservation;
  beforeAll(async () => {
    addReservation = renderer
      .create(
        <Provider store={store}>
          <Router>
            <AddReservation />
          </Router>
        </Provider>,
      )
      .toJSON();
  });

  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('should render correctly', () => {
    expect(addReservation).toBeTruthy();
  });

  it('should match the snapshot', () => {
    expect(addReservation).toMatchSnapshot();
  });
});
