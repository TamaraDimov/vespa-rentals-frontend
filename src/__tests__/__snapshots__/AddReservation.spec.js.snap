// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddReservation should match the snapshot 1`] = `
<div
  className="reservation-container"
>
  <form
    className="reservation-form"
    onSubmit={[Function]}
  >
    <label
      className="form-label"
      htmlFor="start_date"
    >
      Start Date:
      <input
        className="form-input"
        id="start_date"
        name="start_date"
        onChange={[Function]}
        type="date"
        value=""
      />
    </label>
    <label
      className="form-label"
      htmlFor="end_date"
    >
      End Date:
      <input
        className="form-input"
        id="end_date"
        name="end_date"
        onChange={[Function]}
        type="date"
        value=""
      />
    </label>
    <label
      className="form-label"
      htmlFor="city"
    >
      City:
      <select
        className="form-select"
        id="city"
        name="city"
        onChange={[Function]}
        value=""
      >
        <option
          value=""
        >
          Select a city
        </option>
        <option
          value="New York"
        >
          New York
        </option>
        <option
          value="Los Angeles"
        >
          Los Angeles
        </option>
        <option
          value="Chicago"
        >
          Chicago
        </option>
        <option
          value="Houston"
        >
          Houston
        </option>
        <option
          value="Philadelphia"
        >
          Philadelphia
        </option>
      </select>
    </label>
    <label
      className="form-label"
      htmlFor="motorcycle_id"
    >
      Motorcycle:
      <select
        className="form-select"
        id="motorcycle_id"
        name="motorcycle_id"
        onChange={[Function]}
        value=""
      >
        <option
          value=""
        >
          Select a motorcycle
        </option>
        <option
          value={1}
        >
          Honda CBR600RR
        </option>
        <option
          value={2}
        >
          Yamaha YZF-R6
        </option>
      </select>
    </label>
    <button
      disabled={false}
      type="submit"
    >
      Add Reservation
    </button>
  </form>
</div>
`;
